plugins/IPBFormatter/class.ipbformatter.plugin.php

```diff
@@ -141,22 +141,71 @@ public function nbbc() {
         return $this->_NBBC;
     }
 
+    /**
+     * Build an array of attachment records from the Media table, using the MediaID as each record's index.
+     *
+     * @return array|null
+     */
     public function media() {
         if ($this->_Media === null) {
-            try {
-                $i = Gdn::pluginManager()->getPluginInstance('FileUploadPlugin', Gdn_PluginManager::ACCESS_CLASSNAME);
-                $m = $i->mediaCache();
-            } catch (Exception $ex) {
-                $m = array();
+            // Set _Media to a non-null value, so we only do this once.
+            $this->_Media = array();
+
+            // Fire up a basic model, configured for the Media table
+            $mediaModel = new Gdn_Model('Media');
+
+            // Grab a reference to the instance of the current controller
+            $controller = Gdn::Controller();
+
+            // Grab the current discussion ID from the current controller
+            $discussionID = $controller->data('Discussion.DiscussionID');
+
+            // Grab comment data set from the current controller and verify it's populated
+            $comments = $controller->Data('Comments');
+            $commentIDs = array();
+            if ($comments instanceof Gdn_DataSet && $comments->NumRows()) {
+                // Build a collection of comment IDs
+                while ($currentComment = $comments->NextRow()) {
+                    $commentIDs[] = $currentComment->CommentID;
+                }
+                unset($currentComment);
             }
 
-            $media = array();
-            foreach ($m as $key => $data) {
-                foreach ($data as $row) {
-                    $media[$row->MediaID] = $row;
+            /**
+             * Select all media records with a ForeignID matching the discussion ID and a
+             * ForeignTable value of "discussion"
+             */
+            $mediaStatement = $mediaModel->SQL
+                ->Select('m.*')
+                ->From('Media m')
+                ->BeginWhereGroup()
+                ->Where('m.ForeignID', $discussionID)
+                ->Where('m.ForeignTable', 'discussion')
+                ->EndWhereGroup();
+
+            // If we have any comment IDs, find attachments related to them too
+            if (!empty($commentIDs)) {
```

Comment:
Are you sure this is an appropriate query construction for performance? I have a strong suspicion this would perform far better as 2 separate queries and join the datasets afterward.